fix: Adding the fatal error for ipv6 cilium config on a single stack node

vipul-21 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Cilium pods on a single  stack cluster were failing when ipv6 was enabled. The expectation is that ipv4 should be working even if there are error in ipv6.

Fixes: #28909

jschwinger233 reviewed (commented):
I'm not sure about "The expectation is that ipv4 should be working even if there are errors in getting ipv6 ips". Do you mean we expect cilium to work without IPv6 even if IPv6 is enabled by user when DirectRoutingDevice for v6 can't be found?

jschwinger233 reviewed (changes requested):
Thanks! Would you mind squashing two commits into one? The first commit turns fatal to warning, while the second one reverts the change. It could be confusing for git-bisect in future.

ti-mo commented on pkg/datapath/linux/config/config.go:
Please refrain from making stylistic changes like these, it causes unnecessary churn. If gofmt doesn't complain, it's good enough. :)

vipul-21 committed:
fix: Adding ipv6 check on the node init function

Cilium pods on a single stack cluster were failing when ipv6 was enabled.
The change would make sure that if the underlying node does not have ipv6
configuration enabled it would fail even if KPR is disabled.Earlier, if KPR
is enabled, the node would not see a fatal error.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

pippolo84 commented on daemon/k8s/init.go:
```diff
@@ -128,6 +128,18 @@ func WaitForNodeInformation(ctx context.Context, log logrus.FieldLogger, localNo
+					logfields.NodeName:         n.Name,
+					logfields.IPAddr + ".ipv4": nodeIP4,
+					logfields.IPAddr + ".ipv6": nodeIP6,
+				}).Fatal("No IPv6 support on node as ipv6 address is nil")
```
I don't think you need to `Fatal` here. Using `Error` for logging and then returning an error is enough: the agent will be stopped anyway because of the failed hive start phase.

vipul-21 replied:
Sure, let me add that.

pippolo84 commented on daemon/k8s/init.go:
```diff
@@ -128,6 +128,18 @@ func WaitForNodeInformation(ctx context.Context, log logrus.FieldLogger, localNo
 
+		// If the host does not have an IPv6 address, return an error
+		if option.Config.EnableIPv6 {
+			if nodeIP6 == nil {
```
Nit: might as well be `if option.Config.EnableIPv6 && nodeIP6 == nil { ...`

pippolo84 reviewed (changes requested):
Thanks for the PR!
I think the approach is coherent with https://github.com/cilium/cilium/issues/28909#issuecomment-1823325607. I left a couple of suggestions inline.

pippolo84 reviewed (approved):
Thanks! :rocket:
